libbeat/outputs/kafka/config.go

```diff
@@ -272,7 +283,11 @@ func newSaramaConfig(log *logp.Logger, config *kafkaConfig) (*sarama.Config, err
 		retryMax = 1000
 	}
 	k.Producer.Retry.Max = retryMax
-	// TODO: k.Producer.Retry.Backoff = ?
+	k.Producer.Retry.BackoffFunc = func(_, _ int) time.Duration {
+		d := b.WaitDuration()
+		log.Infof("backing off for %v (init: %v, max: %v)", d, config.Backoff.Init, config.Backoff.Max)
+		return d
+	}
```

Comment:
Is BackoffFunc called 'globally' in sarama, or per broker, or per partition? Is the function guaranteed to be called by one go-routine only?

Is a single 'backoff' instance good enough, or do we actually need multiple separate ones?

This function and reset are called from different go-routines. AFAIK the backoff instance used is not threadsafe. Given that we call this function potentially from multiple workers in sarama, this implementation might not exactly give us exponential backoff. If a subset of workers is active, the Reset will continue to reset our state, while errors in more than one worker can `skip` one or two backoff states.

Reply:
Hmm, good points @urso. Thanks! 

Let me look into the first set of questions around Sarama's implementation. 

As for the thread safety, I suppose we could make our backoff instance thread safe so interleaved calls to `WaitDuration` and `Reset` are at least operating on a consistent internal state of the backoff instance. But before doing this I think it's useful to know how Sarama is using `BackoffFunc` internally, so let me start with that.

Reply:
The signature for BackoffFunc is: `BackoffFunc func(retries, maxRetries int) time.Duration`. This indicates that we should have some stateless implementation (sarama tracks state):

e.g.
```
maxBackoffRetries := int(math.Ceil(math.Log2(float64(backoff.Max) / float64(backoff.Init))))

BackoffFunc: func(retries, _ int) time.Duration {
  jitter := ...
  if retries >= maxBackoffRetries {
    return backoff.Max + jitter
  }
  return time.Duration(uint64(backoff.Init) * uint64(1 << retries)) + jitter
}
```

Reply:
> Is BackoffFunc called 'globally' in sarama, or per broker, or per partition? 

It appears to be per partition: https://github.com/Shopify/sarama/blob/58123455d1a70c7f438871597d9a1715462e5d1c/async_producer.go#L497

> Is the function guaranteed to be called by one go-routine only?

No, since there could be multiple goroutines in flight for different partitions: https://github.com/Shopify/sarama/blob/58123455d1a70c7f438871597d9a1715462e5d1c/async_producer.go#L489

Given the above, it means the state of the backoff instance would be shared across multiple partition producers, which is not good. Ideally the state would be per partition producer.

Alternatively we could provide a stateless implementation (like the one [you suggested](https://github.com/elastic/beats/pull/17808#discussion_r416558306)). Of course, that means we can't directly reuse the implementations used by other outputs, i.e. `EqualJitterBackoff`. But I don't see a way around that. I think we'd just have to live with a stateless implementation where the state (i.e. `retries` is managed externally and passed into the backoff function) for the kafka output or a stateful implementation for the other outputs.

Reply:
@urso I implemented a stateless backoff function in 0f7bfae01cd266cf9288d1132c1dc7da0f3ad53b per your suggestion but I deliberately made the calculations in it look more like the ones in the `EqualJitterBackoff` `Wait()` method.